### PR TITLE
[dnssd] add api for accessing queries in the dnssd server

### PR DESCRIPTION
**Pull request by gjc13**
This commit allows the platform implementations to access the name and
the type of the queries, thus reducing the memory overhead of
extra bookkeeping.

Depends-On: https://github.com/openthread/ot-br-posix/pull/841

### CODE REVIEW

File: include/openthread/dnssd_server.h [hunk 1]
```diff
@@ -99,6 +99,12 @@ typedef void (*otDnssdQuerySubscribeCallback)(void *aContext, const char *aFullN
  */
 typedef void (*otDnssdQueryUnsubscribeCallback)(void *aContext, const char *aFullName);
 
+/**
+ * This opaque type represents a dns-sd query.
```
Comment (simonlingoogle) [inline include/openthread/dnssd_server.h:103]:
```suggestion
 * This opaque type represents a DNS-SD query.
```

File: include/openthread/dnssd_server.h [hunk 2]
```diff
@@ -128,6 +134,18 @@ typedef struct otDnssdHostInfo
...
+/**
+ * This enumeration specifies a dns-sd query type.
+ *
+ */
+typedef enum
+{
+    OT_DNSSD_QUERY_TYPE_NONE    = 0, ///< Service type unspecified.
+    OT_DNSSD_QUERY_TYPE_BROWSE  = 1, ///< Service type browse service.
+    OT_DNSSD_QUERY_TYPE_RESOLVE = 2, ///< Service type resolve service instance.
+    OT_DNSSD_QUERY_TYPE_HOST    = 3, ///< Service type resolve hostname.
```
Comment (simonlingoogle) [inline include/openthread/dnssd_server.h:146]:
```suggestion
    OT_DNSSD_QUERY_TYPE_RESOLVE_HOST    = 3, ///< Service type resolve hostname.
```
nit

File: include/openthread/dnssd_server.h [hunk 3]
```diff
@@ -178,6 +196,29 @@ void otDnssdQueryHandleDiscoveredServiceInstance(otInstance *                aIn
...
+ * @param[in] aInstance         The OpenThread instance structure.
+ * @param[in] aQuery            The query pointer. Pass NULL to get the first query.
+ *
+ * @returns  A pointer to the query or NULL if no more queries.
+ *
+ */
+const otDnssdQuery *otDnssdGetNextQuery(otInstance *aInstance, const otDnssdQuery *aQuery);
+
+/**
+ * This function aquires the dns-sd query type and name for a specific query.
```
Comment (simonlingoogle) [inline include/openthread/dnssd_server.h:211]:
```suggestion
 * This function aquires the DNS-SD query type and name for a specific query.
```

File: src/core/net/dnssd_server.hpp [hunk 4]
```diff
@@ -188,19 +127,176 @@ class Server : public InstanceLocator, private NonCopyable
...
+     * @param[out]  aNameOutput       The name output buffer, recommended be at least 255 bytes long.
+     * @param[out]  aNameSize         The size of @p aNameOutput
+     *
+     * @returns The dns-sd query type.
+     *
+     */
+    static DnsQueryType GetQueryTypeAndName(const Header & aHeader,
+                                            const Message &aMessage,
+                                            char *         aNameOutput,
+                                            uint16_t       aNameSize);
```
Comment (simonlingoogle) [inline src/core/net/dnssd_server.hpp:290]:
```suggestion
                                            char (&aName)[Name::kMaxNameSize]);
```
Can we still use a pointer to a fixed size array?

File: src/core/api/dns_server_api.cpp [hunk 5]
```diff
@@ -73,4 +75,26 @@ void otDnssdQueryHandleDiscoveredHost(otInstance *aInstance, const char *aHostFu
...
+otDnssdQueryType otDnssdGetQueryTypeAndName(const otDnssdQuery *aQuery, char *aNameOutput, uint16_t aNameSize)
+{
+    otDnssdQueryType        type  = OT_DNSSD_QUERY_TYPE_NONE;
+    const QueryTransaction *query = static_cast<const QueryTransaction *>(aQuery);
+
+    OT_ASSERT(aQuery != nullptr);
+    OT_ASSERT(aNameOutput != nullptr);
+    OT_ASSERT(query->IsValid());
+    type = static_cast<otDnssdQueryType>(Dns::ServiceDiscovery::Server::GetQueryTypeAndName(
+        query->GetResponseHeader(), query->GetResponseMessage(), aNameOutput, aNameSize));
```
Comment (simonlingoogle) [inline src/core/api/dns_server_api.cpp:95]:
```suggestion
    type = static_cast<otDnssdQueryType>(Dns::ServiceDiscovery::Server::GetQueryTypeAndName(query, aNameOutput, aNameSize));
```

So that we don't need to expose `QueryTransaction` publicly.

File: src/core/net/dnssd_server.hpp [hunk 6]
```diff
@@ -188,19 +127,176 @@ class Server : public InstanceLocator, private NonCopyable
...
+        kDnsQueryResolve     = OT_DNSSD_QUERY_TYPE_RESOLVE, ///< Service type resolve service instance.
+        kDnsQueryResolveHost = OT_DNSSD_QUERY_TYPE_HOST,    ///< Service type resolve hostname.
+    };
+
+public:
+    /**
+     * This class contains the compress information for a dns packet.
+     *
+     */
+    class QueryTransaction
```
Comment (simonlingoogle) [inline src/core/net/dnssd_server.hpp:160]: Keep `QueryTransaction` private?

File: src/core/net/dnssd_server.hpp (hunk 4 above)
Comment (gjc13) [reply src/core/net/dnssd_server.hpp:290]: Since this function will be used by OpenThread c apis, I believe it's no longer possible.

File: src/core/net/dnssd_server.hpp (hunk 6 above)
Comment (gjc13) [reply src/core/net/dnssd_server.hpp:160]: Make `QueryTransaction` private is not possible since we need to call `Server::GetNextQuery(const QueryTransaction *)` outside the class anyway. We can use `otDnssdQuery *` type which is in fact `void *` but I believe it's not in line with the OpenThread style.
Another point for making this public is that users can ony get const reference to the queries. Accessing the memories will be harmless.

Comment (simonlingoogle) [reply src/core/net/dnssd_server.hpp:160]: > Making `QueryTransaction` private is not possible since we need to call `Server::GetNextQuery(const QueryTransaction *)` outside the class anyway. We can use `otDnssdQuery *` type which is in fact `void *` but I believe it's not in line with the OpenThread style.

I think it's fine to use `otDnssdQuery *` in Server.
Making `QueryTransaction` minimizes the code change, which can be much smaller than the current change?

File: include/openthread/dnssd_server.h [hunk 7]
```diff
@@ -178,6 +197,29 @@ void otDnssdQueryHandleDiscoveredServiceInstance(otInstance *                aIn
...
+ *
+ */
+const otDnssdQuery *otDnssdGetNextQuery(otInstance *aInstance, const otDnssdQuery *aQuery);
+
+/**
+ * This function aquires the DNS-SD query type and name for a specific query.
+ *
+ * @param[in]   aQuery            The query pointer acquired from `otDnssdGetNextQuery`.
+ * @param[out]  aNameOutput       The name output buffer, recommended be at least 255 bytes long.
+ * @param[out]  aNameSize         The size of @p aNameOutput
```
Comment (simonlingoogle) [inline include/openthread/dnssd_server.h:216]:
```suggestion
```

File: include/openthread/dnssd_server.h [hunk 8]
```diff
@@ -178,6 +197,29 @@ void otDnssdQueryHandleDiscoveredServiceInstance(otInstance *                aIn
...
+ * @returns  A pointer to the query or NULL if no more queries.
+ *
+ */
+const otDnssdQuery *otDnssdGetNextQuery(otInstance *aInstance, const otDnssdQuery *aQuery);
+
+/**
+ * This function aquires the DNS-SD query type and name for a specific query.
+ *
+ * @param[in]   aQuery            The query pointer acquired from `otDnssdGetNextQuery`.
+ * @param[out]  aNameOutput       The name output buffer, recommended be at least 255 bytes long.
```
Comment (Irving-cl) [inline include/openthread/dnssd_server.h:215]:
```suggestion
 * @param[out]  aNameOutput       The name output buffer, which should be `OT_DNS_MAX_NAME_SIZE` bytes long.
```
By `char (*aNameOutput)[OT_DNS_MAX_NAME_SIZE]`, the buffer has to be `OT_DNS_MAX_NAME_SIZE` long, right?

File: src/core/net/dnssd_server.hpp [hunk 9]
```diff
@@ -188,19 +127,174 @@ class Server : public InstanceLocator, private NonCopyable
...
+     */
+    void SetQueryCallbacks(otDnssdQuerySubscribeCallback   aSubscribe,
+                           otDnssdQueryUnsubscribeCallback aUnsubscribe,
+                           void *                          aContext);
+
+    /**
+     * This method notifies a discovered service instance.
+     *
+     * @param[in] aServiceFullName  The null-terminated full service name.
+     * @param[in] aInstanceInfo     A pointer to the discovered service instance information.
```
Comment (Irving-cl) [inline src/core/net/dnssd_server.hpp:252]:
```suggestion
     * @param[in] aInstanceInfo     A reference to the discovered service instance information.
```

File: src/core/net/dnssd_server.hpp [hunk 10]
```diff
@@ -188,19 +127,174 @@ class Server : public InstanceLocator, private NonCopyable
...
+     *
+     */
+    void HandleDiscoveredServiceInstance(const char *aServiceFullName, const otDnssdServiceInstanceInfo &aInstanceInfo);
+
+    /**
+     * This method notifies a discovered host.
+     *
+     * @param[in] aHostFullName     The null-terminated full host name.
+     * @param[in] aHostInfo         A pointer to the discovered host information.
```
Comment (Irving-cl) [inline src/core/net/dnssd_server.hpp:261]:
```suggestion
     * @param[in] aHostInfo         A reference to the discovered host information.
```

File: src/core/net/dnssd_server.hpp [hunk 11]
```diff
@@ -188,19 +127,174 @@ class Server : public InstanceLocator, private NonCopyable
...
+     *
+     */
+    void HandleDiscoveredHost(const char *aHostFullName, const otDnssdHostInfo &aHostInfo);
+
+    /**
+     * This function aquires the next query in the server.
+     *
+     * @param[in] aQuery            The query pointer. Pass NULL to get the first query.
```
Comment (Irving-cl) [inline src/core/net/dnssd_server.hpp:269]:
```suggestion
     * @param[in] aQuery            The query pointer. Pass nullptr to get the first query.
```
Since we're in hpp.

File: src/core/net/dnssd_server.hpp [hunk 12]
```diff
@@ -188,19 +127,174 @@ class Server : public InstanceLocator, private NonCopyable
...
+     *
+     * @returns  A pointer to the query or NULL if no more queries.
```
Comment (Irving-cl) [inline src/core/net/dnssd_server.hpp:271]:
```suggestion
     * @returns  A pointer to the query or nullptr if no more queries.
```

File: src/core/net/dnssd_server.hpp [hunk 13]
```diff
@@ -107,27 +119,36 @@ class Server : public InstanceLocator, private NonCopyable
...
+    /**
+     * This function aquires the dns-sd query type and name for a specific query.
+     *
+     * @param[in]   aQuery            The query pointer.
+     * @param[out]  aNameOutput       The name output buffer.
+     *
+     * @returns The dns-sd query type.
+     *
+     */
+    static DnsQueryType GetQueryTypeAndName(const otDnssdQuery *aQuery, char (&aName)[Name::kMaxNameSize]);
```
Comment (simonlingoogle) [inline src/core/net/dnssd_server.hpp:146]:
```suggestion
    static DnsQueryType GetQueryTypeAndName(const otDnssdQuery &aQuery, char (&aName)[Name::kMaxNameSize]);
```
nit: use reference

**Comment by gjc13** [comment]
@Irving-cl @simonlingoogle All comments resolved, PTAL.

File: src/core/net/dnssd_server.hpp [hunk 14]
```diff
@@ -188,19 +214,26 @@ class Server : public InstanceLocator, private NonCopyable
             }
         }
 
-    private:
         static bool MatchCompressedName(const Message &aMessage, uint16_t aOffset, const char *aName)
```
Comment (simonlingoogle) [inline src/core/net/dnssd_server.hpp:217]: Do we still need to make `MatchCompressedName` public?

File: src/core/net/dnssd_server.hpp [hunk 15]
```diff
@@ -232,47 +265,66 @@ class Server : public InstanceLocator, private NonCopyable
...
+         */
+        const Message &GetResponseMessage(void) const { return *mResponseMessage; }
+
+        /**
+         * This method checks whether the query is valid.
+         *
+         * @returns Whether the query is valid.
+         *
+         */
+        bool IsValid(void) const { return mResponseMessage != nullptr; }
```
Comment (simonlingoogle) [inline src/core/net/dnssd_server.hpp:304]: Do these methods need to be public?

File: src/core/net/dnssd_server.hpp [hunk 16]
```diff
@@ -323,6 +375,9 @@ class Server : public InstanceLocator, private NonCopyable
                                          Message &               aMessage,
                                          const Ip6::MessageInfo &aMessageInfo,
                                          Ip6::Udp::Socket &      aSocket);
+    static DnsQueryType     GetQueryTypeAndName(const Header & aHeader,
```
Comment (simonlingoogle) [inline src/core/net/dnssd_server.hpp:378]: seems no change to this method. Why move location? Should we minimize unrelated change?

**Review by simonlingoogle — APPROVED** [review]
LGTM :+1: 

nit: Better keep only related changes, but there seems to be several unnecessary changes.

**Review by Irving-cl — APPROVED** [review]
LGTM :+1:

File: src/core/net/dnssd_server.hpp (hunk 13 above)
Comment (gjc13) [reply src/core/net/dnssd_server.hpp:146]: Since `otDnssdQuery` is in fact `void`, it's not possible to have a void reference.

**Comment by gjc13** [comment]
Unrelated changes removed.

**Comment by gjc13** [comment]
@jwhui, all comments resolved.

**Comment by jwhui** [comment]
> @jwhui, all comments resolved.

Thanks. Still waiting on @simonlingoogle 's review of https://github.com/openthread/ot-br-posix/pull/841.

**Review by jwhui — APPROVED** [review]

File: include/openthread/dnssd_server.h (hunk 8 above)
Comment (wgtdkp) [reply include/openthread/dnssd_server.h]: I don't think so, array in C laungage is just pointer, there is no length properties of a array type.
